### ISSUE TITLE
Orquesta design: task failures in branches

### ISSUE THREAD

**Issue by emptywee** [StackStorm/st2#5054]
This is a continuation of https://github.com/StackStorm/st2/issues/5031

Currently Orquesta's design for failures in branches is to terminate the whole workflow as soon as possible once one branch finishes in a failure (in case if one branch finishes with success, the engine will keep other branches running). To me, it's a huge drawback and it's not consistent with mistral's behavior from which we're supposed to migrate from. Also, it's counter-intuitive as you expect the branches running in parallel to be independent from each other. 

However, I understand there could be circumstances where you'd want to terminate your workflow entirely if one of the tasks in some branch fails. It's probably a good thing to have as an option, so maybe having a special transition named `terminate` or something like that would be a good choice here.

I invite all concerned parties to discuss this topic and encourage everyone to chime in. 

P.S. It reminded me of the time when mistral did not have the `publish-on-error` feature, where I'd have to handle all failed tasks with additional `core.noop` tasks to publish what error happened exactly so that it would be available in the final output...

**Comment by armab** [comment]
Just yesterday example from the real world:
Our TravisCI has `5` branches.
Based on: https://travis-ci.org/github/StackStorm/st2/builds/733079011 `1` failed and `4` succeeded.
All those branches finished and reported their status, failed job didn't block others from the completion and it's expected them to be parallel/independent.

I think that's the most used in a wild and expected workflow behavior from the user's side and can understand @emptywee frustration. I bet we'll see these cases of misunderstanding from the community in future as well.

@m4dcoder What is the way to achieve that behavior with the Orquesta workflow engine? `Join: all`?

**Comment by emptywee** [comment]
@armab is the link correct? That build has all 5 succeeded. But https://travis-ci.org/github/StackStorm/st2/builds/733441608 has 1 failed and 4 succeeded, just like you mentioned.

**Comment by emptywee** [comment]
And it is not always that simple with joins in orquesta...

**Comment by cognifloyd** [comment]
As the semantics of the `fail` command are already defined for orquesta workflows, maybe we need a new engine command that more closely aligns with the mistral semantics.

[Current engine commands](https://docs.stackstorm.com/orquesta/languages/orquesta.html#engine-commands):

> Command | Description
> -- | --
> continue | Default value when do is not specified. The workflow engine will not alter the previous task state and will continue to conduct the workflow execution. If the previous task state is one of the failure states, the conductor will continue and fail the workflow execution.
> fail | The workflow engine will fail the workflow execution.
> noop | The workflow engine will perform no operation given previous task state. If the previous task state is one of the failure states, the conductcor will ignore the task failure and assume a remediation has occurred.
> retry | The workflow engine will retry the task up to 3 times with no delay.

The description of `fail` on this page needs to mention that it triggers the fail-fast behavior.

So maybe something like this:

Command | Description
-- | --
fail | The workflow engine will fail the workflow execution. The workflow will exit immediately once all active tasks complete, aka "fail fast".
wait-and-fail | The workflow engine will wait for all active branches to finish and then exit the workflow with a failed state. Warning: This will increase workflow run-times and might delay error reporting. So, we recommend using `fail` whenever possible. That said, this provides semantics that more closely align with the Mistral workflow behavior, so it may facilitate conversion from Mistral to Orquesta workflows.

As this is something that I'm pretty sure @m4dcoder will want to discourage using, a slightly longer name, like `wait-and-fail` should make it clear that this is not the recommended approach, but if it simplifies a workflow's design, go for it.

**Comment by m4dcoder** [comment]
The option (term_on_fail or terminate on failure) that I'm thinking of is more of a runner parameter and a st2 global config option.  The examples are listed below.  User can either specify the option per workflow execution or globally configured for all workflow execution.

## runner parameter (affects per workflow)
For example, `st2 run mypack.myworkflow term_on_fail=False` 

## st2 global config (affects all workflow)
```
[workflow_engine]
term_on_fail = False
```

When this option is false, the workflow execution will stay running and workflow engine will continue to schedule task execution until there are no more next tasks. The final status of the workflow is "failed" and output is determined accordingly.

**Comment by cognifloyd** [comment]
Hmm. And then the workflow's st2 metadata could set `term_on_fail` to false and perhaps even immutable depending on the design.

I think that could work. But, then there is no indication in the workflow file that it is designed to use `term_on_fail=False`. What about an option in the workflow itself to set the default for the given workflow?

**Comment by armab** [comment]
I agree that workflows with the same yaml code behaving differently depending on st2.conf setting would lead us to issues. So yeah, there should be something within workflow spec to indicate the logic.

**Comment by m4dcoder** [comment]
Sure, we can scratch the above and do something like the following in the workflow definition.  

```
version: 1.0

description: ...

runtime:
  term_on_fail: false

tasks:
 ...

```

**Comment by nmaludy** [comment]
Had another use case for this. We use StackStorm to backup the various database in our environment like so:

```

  db_backup:
    action: encore.bolt_plan
    with:
      items: "{{ ctx().database_vms }}"
      concurrency: "{{ ctx().backup_concurrency }}"
    input:
      server_ip_address: "{{ item().name }}"
      bolt_retry_count: 0
      bolt_timeout: "{{ ctx().backup_timeout }}"
      os_type: linux
      plan: encore_rp::db_backup
      params:
        databases: "{{ item().value.split(',') }}"
        db_username: "{{ ctx().db_username }}"
        db_password: "{{ ctx().db_password }}"
```

I _really_ want to ensure that all of those run, regardless if any individual one succeeds or fails, they all at least need to execute. We can gather up the failures at the end and make tickets for a human to go fix, but i can't afford for one database to not get backed up.

**Comment by cognifloyd** [comment]
Bye bye stale bot
